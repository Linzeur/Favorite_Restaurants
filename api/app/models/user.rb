class User < ApplicationRecord

  validates :name, presence: true
  validates :email, format: { with: /[a-zA-Z0-9]*@[a-zA-Z]*.[a-zA-Z0-9]*/, message: "Bad format"  },
                    uniqueness: true

  has_secure_password
  has_secure_token
  has_secure_token :reset_digest

  has_many :locations
  has_many :favorite_restaurants

  def self.valid_login?(email, password)
    user = find_by(email: email)
    user if user&.authenticate(password)
  end

  def invalidate_token
    update(token: nil)
  end
end
