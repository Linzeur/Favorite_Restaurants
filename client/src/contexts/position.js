import React from "react";

const PositionContext = React.createContext();

function PositionProvider(props) {
  const [position, setPosition] = React.useState([0, 0, false]);

  React.useEffect(() => {
    const watchId = navigator.geolocation.watchPosition(({ coords }) => {
      setPosition([coords.latitude, coords.longitude, true]);
      sessionStorage.setItem(
        "lastCurrentPosition",
        JSON.stringify({ lat: coords.latitude, lng: coords.longitude })
      );
    });
    return () => navigator.geolocation.clearWatch(watchId);
  }, [setPosition]);

  const value = {
    latitude: position[0],
    longitude: position[1],
    callAPI: position[2],
    onChange: (lat, lng, callAPI) => {
      setPosition([lat, lng, callAPI]);
    }
  };

  return <PositionContext.Provider value={value} {...props} />;
}

export { PositionProvider, PositionContext };
