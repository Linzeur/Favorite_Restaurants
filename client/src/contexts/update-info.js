import React from "react";

const UpdateListContext = React.createContext();

function UpdateListProvider(props) {
  const [updateList, setUpdateList] = React.useState({
    locations: 0,
    favorites: 0
  });

  const value = {
    locations: updateList.locations,
    favorites: updateList.favorites,
    onChange: key => {
      setUpdateList(prevState => {
        return { ...prevState, [key]: prevState[key] + 1 };
      });
    }
  };

  return <UpdateListContext.Provider value={value} {...props} />;
}

export { UpdateListProvider, UpdateListContext };
