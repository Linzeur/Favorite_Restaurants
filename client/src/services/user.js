import { apiUrl } from "../util";

const API_USERS = `${apiUrl}users`;

async function createUser(userData) {
  const response = await fetch(API_USERS, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(userData)
  });

  if (!response.ok) {
    const { errors } = await response.json();
    throw new Error(errors.message);
  }

  return response.json();
}

export { createUser };
