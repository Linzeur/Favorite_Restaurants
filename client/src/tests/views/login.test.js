import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";

import Login from "../../views/login";
import { UserProvider } from "../../contexts/user";

test("Login component view", () => {
  const { asFragment } = render(
    <UserProvider>
      <Login />
    </UserProvider>
  );

  expect(asFragment()).toMatchSnapshot();
});

const userLogged = { name: "Usuario Test", email: "test@test.com" };

test("log in a valid user", async () => {
  fetch.mockResponseOnce(JSON.stringify(userLogged));
  const { getByLabelText, getByRole } = render(
    <UserProvider>
      <Login />
    </UserProvider>
  );

  const inputEmail = getByLabelText("Enter your email");
  const inputPassword = getByLabelText("Enter your password");
  const buttonLogin = getByRole("button");

  fireEvent.change(inputEmail, {
    target: { value: "test@test.com" }
  });

  fireEvent.change(inputPassword, {
    target: { value: "123456" }
  });

  fireEvent.submit(buttonLogin);

  await wait(() => {});

  const user = JSON.parse(localStorage.getItem("user"));
  expect(user.name).toBe(userLogged.name);
  expect(user.email).toBe(userLogged.email);
});
