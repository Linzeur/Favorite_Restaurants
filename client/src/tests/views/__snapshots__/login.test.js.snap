// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component view 1`] = `
<DocumentFragment>
  <div
    class="css-1s2t2fo-Login"
  >
    <form
      class="css-fl0flg-Login"
    >
      <h1
        class="css-q6ww2i-Login"
      >
        LOGIN
      </h1>
      <label
        class="css-jidqlr-Login"
        for="email"
      >
        Email
      </label>
      <input
        aria-label="Enter your email"
        class="css-ulpja2-Login"
        id="email"
        name="email"
        placeholder="Enter your email"
        required=""
        type="email"
        value=""
      />
      <br />
      <label
        class="css-jidqlr-Login"
        for="password"
      >
        Password
      </label>
      <input
        aria-label="Enter your password"
        autocomplete="off"
        class="css-ulpja2-Login"
        id="password"
        name="password"
        placeholder="Enter your password"
        required=""
        type="password"
        value=""
      />
      <br />
      <button
        aria-label="Sign In User"
        class="css-1vdzh6a-Login"
        type="submit"
      >
        Log In
      </button>
      <br />
      <label
        class="css-1fdx9wf-Login"
      >
        <a
          href="/signup"
        >
          Go to Sign up
        </a>
      </label>
      <br />
    </form>
  </div>
</DocumentFragment>
`;
